### ISSUE TITLE
Safe helm updates

### ISSUE THREAD

**Issue by rothgar** [aws-controllers-k8s/community#698]
If a user has a stable chart installed and runs `helm upgrade` it will automatically upgrade them to the latest published chart (stable or unstable).

This could cause breaking changes or unintended stability issues. I have found 3 possible solutions to make this better for users.

1. Create separate repos for helm releases (stable and beta?) and let users select which updates they want to subscribe to based on the repo they add in helm. This would prevent stable -> beta changes but would not fix the problem of major releases from stable -> stable if there are breaking changes (e.g. v1 -> v2)
2. We can use deprecation notifications to block upgrades until users change configuration to meet the new requirements. This is the [approach Gitlab uses](https://docs.gitlab.com/charts/architecture/decisions.html#breaking-changes-via-deprecation). It's nice that this doesn't require two repos and requires manual updates from a user to make sure they deploy the correct configuration with each chart if there are breaking changes.
3. Ideally there would be a way in helm to lock an installed version. This isn't possible today but it could work similar to how apt and yum work by specifying a version inside a config and each time `helm upgrade` or `helm install` is run it checks that version. This isn't possible today but I think it's a needed feature in helm in general. I'm going to work work with helm developers to suggest this feature but there's no timeline for when it could be available.

With those options I think number 2 is the easiest and most actionable right now. We don't have any charts that have breaking changes but we should document how to do that in our release documents and mention it in our user documents.

If that sounds like a good plan please :+1: this and I can work on doc changes.

**Comment by rothgar** [comment]
As an example of what a deprecation message looks like you can see what it looks like to install a chart without required parameters. It's not the most user friendly output but it can have free form text to specify what needs to be included.

```
WARNING: This chart is deprecated
NAME: autoscaler
LAST DEPLOYED: Tue Feb 23 17:58:07 2021
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
##############################################################################
####   ERROR: You must specify values for either                          ####
####   autoDiscovery.clusterName or autoscalingGroups[]                   ####
##############################################################################

The deployment and pod will not be created and the installation is not functional
See README:
  open https://github.com/kubernetes/charts/tree/master/stable/cluster-autoscaler
```

**Comment by jaypipes** [comment]
> If a user has a stable chart installed and runs `helm upgrade` it will automatically upgrade them to the latest published chart (stable or unstable).

Assume the user has `helm install`'d an ACK service controller chart with a `chart:s3-v1-stable` tag (note, we have the major version included in the Helm Chart stable tag now; there is no `-stable` tag suffix without a major version). Let's say this Helm Chart's Deployment image is `controller:s3-v1.2.8`

Now assume that we release a v2 series S3 ACK service controller and eventually release a Helm Chart that installs a stable image of this v2 series controller with a `chart:s3-v2-stable` tag. This Helm Chart's Deployment image is `controller:s3-v2.0.3`.

If the user does a raw `helm upgrade`, are you saying that Helm will install the `chart:s3-v2-stable` chart? If that is the case, that is super bad. :(

> This could cause breaking changes or unintended stability issues. I have found 3 possible solutions to make this better for users.
> 
>     1. Create separate repos for helm releases (stable and beta?) and let users select which updates they want to subscribe to based on the repo they add in helm. This would prevent stable -> beta changes but would not fix the problem of major releases from stable -> stable if there are breaking changes (e.g. v1 -> v2)

Yeah, this solution is icky, IMHO.

>     2. We can use deprecation notifications to block upgrades until users change configuration to meet the new requirements. This is the [approach Gitlab uses](https://docs.gitlab.com/charts/architecture/decisions.html#breaking-changes-via-deprecation). It's nice that this doesn't require two repos and requires manual updates from a user to make sure they deploy the correct configuration with each chart if there are breaking changes.

Slightly less icky, yes.

>     3. Ideally there would be a way in helm to lock an installed version. This isn't possible today but it could work similar to how apt and yum work by specifying a version inside a config and each time `helm upgrade` or `helm install` is run it checks that version. This isn't possible today but I think it's a needed feature in helm in general. I'm going to work work with helm developers to suggest this feature but there's no timeline for when it could be available.

Awesome. Keep us in the loop as to how that conversation with the Helm devs goes. I'm obviously very interested in the outcome.

> With those options I think number 2 is the easiest and most actionable right now. We don't have any charts that have breaking changes but we should document how to do that in our release documents and mention it in our user documents.

Agreed. I'll be working on the release documentation (the contributor-focused docs, not the end-user docs) in the next couple days and will add information about stable releases.

> If that sounds like a good plan please +1 this and I can work on doc changes.

:+1: from me, @rothgar

**Comment by rothgar** [comment]
>If the user does a raw helm upgrade, are you saying that Helm will install the chart:s3-v2-stable chart?
Yes. `helm upgrade` will automatically pull in the latest published release and currently there is no way to prevent that unless the user specifies the specific version they want.

**Comment by jaypipes** [comment]
> > If the user does a raw helm upgrade, are you saying that Helm will install the chart:s3-v2-stable chart?
> 
> Yes. `helm upgrade` will automatically pull in the latest published release and currently there is no way to prevent that unless the user specifies the specific version they want.

ewww. Yes, we should definitely work with the Helm community on that. Seems like a big problem.

**Comment by rothgar** [comment]
Following the process for a [Helm Improvement Proposal](https://github.com/helm/community/blob/master/hips/hip-0001.md) (HIP) I sent an email to the mailing list [here](https://lists.cncf.io/g/cncf-helm/topic/hip_helm_lock/80907000?p=,,,20,0,0,0::recentpostdate%2Fsticky,,,20,2,0,80907000) before I create a formal proposal.

I'm guessing this feature review, implementation, and release is going to take a while so we should try our best not to make breaking changes without having deprecation notifications.
